### PR TITLE
Stars: Add dashboard_uid and org_id to table

### PR DESCRIPTION
**Pull request by stephaniehingtgen**
**What is this feature?**

This PR adds in `dashboard_uid` and `org_id` to the `star` table, to help us in our transition away from using dashboard IDs. It then switches any internal usage of dashboard IDs on the star table to begin using dashboard uids instead.

In addition to running the migration to fill in this data during the sql migrations, it also runs it at startup, in case a grafana instance downgrades, adds a star, and then upgrades. This ensures dashboard_uid and org_id are always present.

Finally, it updates the docs to use the /uid endpoint for stars. The /id endpoint is [deprecated](https://github.com/grafana/grafana/blob/main/pkg/services/star/api/api.go#L90) and has been for 2 years.

It also adds the `updated` column while we are at it :)

**Which issue(s) does this PR fix?**:

Relates to https://github.com/grafana/app-platform-wg/issues/177

### CODE REVIEW

File: pkg/services/sqlstore/migrations/migrations.go [hunk 1]
```diff
@@ -157,4 +157,17 @@ func addStarMigrations(mg *Migrator) {
...
+	mg.AddMigration("Add column dashboard_uid in star", NewAddColumnMigration(starV1, &Column{
+		Name: "dashboard_uid", Type: DB_Text, Nullable: true,
+	}))
+	mg.AddMigration("Add column org_id in star", NewAddColumnMigration(starV1, &Column{
+		Name: "org_id", Type: DB_BigInt, Nullable: false, Default: "1",
+	}))
+	mg.AddMigration(
+		"Update dashboard_uid and org_id value in star",
+		NewRawSQLMigration("").
+			SQLite("UPDATE star SET dashboard_uid = (SELECT uid FROM dashboard WHERE dashboard.id = star.dashboard_id), org_id = (SELECT org_id FROM dashboard WHERE dashboard.id = star.dashboard_id);").
```
Comment (ryantxu) [inline pkg/services/sqlstore/migrations/migrations.go:169]: to support backwards migratinos (and then later coming forwar) this logic needs to be move to a background startup/cleanup job like we have for:

https://github.com/grafana/grafana-enterprise/blob/v11.3.0/src/pkg/extensions/report/database/database_mig.go#L306

File: pkg/services/sqlstore/migrations/migrations.go [hunk 2]
```diff
@@ -157,4 +157,17 @@ func addStarMigrations(mg *Migrator) {
 
 	mg.AddMigration("create star table", NewAddTableMigration(starV1))
 	mg.AddMigration("add unique index star.user_id_dashboard_id", NewAddIndexMigration(starV1, starV1.Indices[0]))
+	mg.AddMigration("Add column dashboard_uid in star", NewAddColumnMigration(starV1, &Column{
+		Name: "dashboard_uid", Type: DB_Text, Nullable: true,
+	}))
+	mg.AddMigration("Add column org_id in star", NewAddColumnMigration(starV1, &Column{
+		Name: "org_id", Type: DB_BigInt, Nullable: false, Default: "1",
+	}))
+	mg.AddMigration(
```
Comment (ryantxu) [inline pkg/services/sqlstore/migrations/migrations.go:166]: while we are hear, can we add an updated time field?  then we can use if for the rV

File: pkg/services/sqlstore/migrations/migrations.go [hunk 3]
```diff
@@ -141,20 +141,3 @@ func (oss *OSSMigrations) AddMigration(mg *Migrator) {
 
 	accesscontrol.AddReceiverCreateScopeMigration(mg)
 }
-
-func addStarMigrations(mg *Migrator) {
```
Comment (stephaniehingtgen) [inline pkg/services/sqlstore/migrations/migrations.go:145]: Moved this to a separate file

Comment (stephaniehingtgen) [inline pkg/services/sqlstore/migrations/migrations.go:145]: Moved this to a separate file

File: pkg/services/sqlstore/migrations/migrations.go (hunk 2 above)
Comment (stephaniehingtgen) [reply pkg/services/sqlstore/migrations/migrations.go]: added :)

File: pkg/services/sqlstore/migrations/migrations.go [hunk 4]
```diff
@@ -157,4 +157,17 @@ func addStarMigrations(mg *Migrator) {
 
 	mg.AddMigration("create star table", NewAddTableMigration(starV1))
 	mg.AddMigration("add unique index star.user_id_dashboard_id", NewAddIndexMigration(starV1, starV1.Indices[0]))
+	mg.AddMigration("Add column dashboard_uid in star", NewAddColumnMigration(starV1, &Column{
+		Name: "dashboard_uid", Type: DB_Text, Nullable: true,
+	}))
+	mg.AddMigration("Add column org_id in star", NewAddColumnMigration(starV1, &Column{
+		Name: "org_id", Type: DB_BigInt, Nullable: false, Default: "1",
```
[the remark this answers is not in the document]
Comment (stephaniehingtgen) [reply pkg/services/sqlstore/migrations/migrations.go]: Good catch! Moved them to nullable

File: pkg/services/sqlstore/migrations/migrations.go (hunk 1 above)
Comment (stephaniehingtgen) [reply pkg/services/sqlstore/migrations/migrations.go]: Done!

**Review by papagian — APPROVED** [review]
I have tested against all the backends and it works as expected 🎉 
I think we need to introduce an additional index.

File: pkg/services/star/starimpl/xorm_store.go [hunk 5]
```diff
@@ -14,9 +14,22 @@ type sqlStore struct {
 func (s *sqlStore) Get(ctx context.Context, query *star.IsStarredByUserQuery) (bool, error) {
 	var isStarred bool
 	err := s.db.WithDbSession(ctx, func(sess *db.Session) error {
+		if query.DashboardUID != "" && query.OrgID != 0 {
+			rawSQL := "SELECT 1 from star where user_id=? and dashboard_uid=? and org_id=?"
```
[the remark this answers is not in the document]
Comment (stephaniehingtgen) [reply pkg/services/star/starimpl/xorm_store.go:18]: Good idea! Added :)

File: pkg/services/sqlstore/migrations/migrations.go [hunk 6]
```diff
@@ -32,9 +32,9 @@ func (oss *OSSMigrations) AddMigration(mg *Migrator) {
 	mg.AddCreateMigration()
 	addUserMigrations(mg)
 	addTempUserMigrations(mg)
-	addStarMigrations(mg)
 	addOrgMigrations(mg)
 	addDashboardMigration(mg) // Do NOT add more migrations to this function.
+	addStarMigrations(mg)
```
[the remark this answers is not in the document]
Comment (stephaniehingtgen) [reply pkg/services/sqlstore/migrations/migrations.go]: Ah, good call! Updated :)

File: docs/sources/developers/http_api/user.md [hunk 7]
```diff
@@ -570,14 +570,14 @@ Content-Type: application/json
 
 ## Star a dashboard
 
-`POST /api/user/stars/dashboard/:dashboardId`
+`POST /api/user/stars/dashboard/uid/:uid`
```
Comment (ryantxu) [inline docs/sources/developers/http_api/user.md:573]: 🎉

File: pkg/services/star/model.go [hunk 8]
```diff
@@ -1,37 +1,48 @@
...
 
 var ErrCommandValidationFailed = errors.New("command missing required fields")
 
 type Star struct {
-	ID          int64 `xorm:"pk autoincr 'id'" db:"id"`
-	UserID      int64 `xorm:"user_id" db:"user_id"`
-	DashboardID int64 `xorm:"dashboard_id" db:"dashboard_id"`
+	ID           int64     `xorm:"pk autoincr 'id'" db:"id"`
+	UserID       int64     `xorm:"user_id" db:"user_id"`
+	DashboardID  int64     `xorm:"dashboard_id" db:"dashboard_id"`
```
Comment (ryantxu) [inline pkg/services/star/model.go:13]:
```suggestion
    // Deprecated: use DashboardUID
	DashboardID  int64     `xorm:"dashboard_id" db:"dashboard_id"`
```

File: pkg/services/star/model.go [hunk 9]
```diff
@@ -1,37 +1,48 @@
...
 }
 
 // ----------------------
 // COMMANDS
 
 type StarDashboardCommand struct {
-	UserID      int64 `xorm:"user_id"`
-	DashboardID int64 `xorm:"dashboard_id"`
+	UserID       int64     `xorm:"user_id"`
+	DashboardID  int64     `xorm:"dashboard_id"`
```
Comment (ryantxu) [inline pkg/services/star/model.go:24]:
```suggestion
    // Deprecated: use DashboardUID
	DashboardID  int64     `xorm:"dashboard_id"`
```

File: pkg/services/star/model.go [hunk 10]
```diff
@@ -1,37 +1,48 @@
...
+	DashboardUID string    `xorm:"dashboard_uid"`
+	OrgID        int64     `xorm:"org_id"`
+	Updated      time.Time `xorm:"updated"`
 }
 
 func (cmd *StarDashboardCommand) Validate() error {
-	if cmd.DashboardID == 0 || cmd.UserID == 0 {
+	if (cmd.DashboardID == 0 && cmd.DashboardUID == "" && cmd.OrgID == 0) || cmd.UserID == 0 {
```
Comment (ryantxu) [inline pkg/services/star/model.go:31]:
```suggestion
	// nolint:staticcheck
	if (cmd.DashboardID == 0 && cmd.DashboardUID == "" && cmd.OrgID == 0) || cmd.UserID == 0 {
```

**Comment by ryantxu** [comment]
This is great -- IIUC, it means we don't need to think about a stars resource migration for a while 🤔

**Review by ryantxu — APPROVED** [review]

**Pull request merged**